contracts/src/registry/NameWrapperRegistry.sol

```diff
@@ -0,0 +1,247 @@
+import {EnhancedAccessControl} from "./EnhancedAccessControl.sol";
+
+contract NameWrapperRegistry is PermissionedRegistry, EnhancedAccessControl {
+    bytes32 public constant REGISTRAR_ROLE = keccak256("REGISTRAR_ROLE");
+    bytes32 public constant EMANCIPATOR_ROLE = keccak256("EMANCIPATOR_ROLE");
+    bytes32 public constant EMANCIPATED_OWNER_ROLE = keccak256("EMANCIPATED_OWNER_ROLE");
+    bytes32 public constant RENEW_ROLE = keccak256("RENEW_ROLE");
+
+    error NameAlreadyRegistered(string label);
+    error NameExpired(uint256 tokenId);
+    error CannotReduceExpiration(uint64 oldExpiration, uint64 newExpiration);
+    event NameRenewed(uint256 indexed tokenId, uint64 newExpiration, address renewedBy);
+    error NameTransferLocked(uint256 tokenId);
+
+    /**
+     * @dev Prevents a name from being transferred.
+     *
+     * We use a boolean for this instead of roles due to the added complexity of dealing with 
+     * ERC1155 approved operators if we were to use roles instead.
+     */
+    mapping(bytes32 tokenIdContext => bool locked) private transferLock;
+
+    constructor(IRegistryDatastore _datastore) PermissionedRegistry(_datastore) {
+        _grantRole(ROOT_CONTEXT, DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function uri(uint256 /*tokenId*/ ) public pure override returns (string memory) {
+        return "";
+    }
+
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
+    {
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(oldFlags);
+        if (expires < block.timestamp) {
+            return address(0);
+        }
+        return super.ownerOf(tokenId);
+    }
+
+    function register(string calldata label, address owner, IRegistry registry, uint96 flags, uint64 expires)
+        public
+        onlyRole(ROOT_CONTEXT, REGISTRAR_ROLE)
+        returns (uint256 tokenId)
+    {
+        tokenId = _canonicalTokenId(uint256(keccak256(bytes(label))), flags);
+        flags = (flags & FLAGS_MASK) | (uint96(expires) << 32);
+
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiry = _extractExpiry(oldFlags);
+        if (oldExpiry >= block.timestamp) {
+            revert NameAlreadyRegistered(label);
+        }
+
+        // if there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        datastore.setSubregistry(tokenId, address(registry), flags);
+
+        // registrar has some permissions by default
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _grantRole(tokenIdContext, RENEW_ROLE, msg.sender);
+        _grantRole(tokenIdContext, EMANCIPATOR_ROLE, msg.sender);
```

Comment:
This seems dangerous insofar as changing the registrar won't change who has control over existing names, which may be an unintended gotcha. What about using the admin fallthrough and granting these roles in the root context to the registrar account(s)? This would also make it easier to revoke this right from all domains at once.

Reply:
Good point!

Reply:
Because we want to be able to emancipate one name at a time, no?

Reply:
I think we could still do this, since rights applied to individual resources override those applied globally.